Add vote controller and test.

Murph33 opened the pull request:
Make minor changes to views to accomodate voting and place each quote in
a well.

Senjai commented on app/controllers/quotes_controller.rb:
```diff
@@ -3,6 +3,7 @@ class QuotesController < ApplicationController
 
   def show
     @quote = Quote.find params[:id]
+    @vote = @quote.votes.find_by(user: current_user)
```
Does this need to be in the same commit?

Murph33 replied:
It would break the quote show page without if I'm remembering right.  Maybe I added more than I should have by including view functionality at the same time?  Was trying to imitate the quote controller commit I'd made before.

Senjai commented on app/controllers/votes_controller.rb:
```diff
@@ -0,0 +1,35 @@
+class VotesController < ApplicationController
+  before_action :authenticate_user!
+
+  def create
```
Each of these actions and their tests can be in their own commits.

Senjai commented on app/models/quote.rb:
```diff
@@ -4,4 +4,10 @@ class Quote < ActiveRecord::Base
 
   has_many :votes, dependent: :destroy
+
+  def score
```
Seperate commit here.

Senjai commented on app/models/quote.rb:
```diff
@@ -4,4 +4,10 @@ class Quote < ActiveRecord::Base
+
+  def score
+    score = 0
+    votes.each { |vote| score += vote.value }
```
`votes.sum(:value)` assuming value is a database column.

Senjai commented on app/views/quotes/index.html.erb:
```diff
@@ -1,3 +1,6 @@
 <% @quotes.each do |quote| %>
-  <%= render quote %>
+  <div class="well">
```
Each of these views can be in seperate commits with the controller actions.

Senjai commented on app/views/votes/_voting.html.erb:
```diff
@@ -0,0 +1,14 @@
+<% if Vote.exists?(quote: quote, user: current_user) %>
+  <%= link_to "I hate it", quote_vote_path(quote, vote, { vote: { value: -2 } } ), method: :patch %>
```
Magic numbers like -2 and -1 don't really mean anything. We should have this be enum'd in some fashion. `VOTE_VALUES[:hate_it]` or the like. This allows you to write validations on how much someone can vote something down or up, and you only have to change it in one place if you want to tweak these values.

Senjai commented:
Added a few comments primarily focused on breaking this up into smaller commits. If you have to say "as well" in your commit message, it can probably be a separate commit ;)

Senjai commented on app/views/votes/_voting.html.erb:
```diff
@@ -0,0 +1,14 @@
+<% if Vote.exists?(quote: quote, user: current_user) %>
```
This is going to be an N+1 query. For every quote on the page we query to see if the user has voted for this quote or not. I'd like to see this be done better, ideally with a single query for each set of results (or page, if pagination is added at some point).

Senjai replied:
Ignoring perf for the quoteboard. We can ignore this one.

Murph33 committed:
Add uniqueness scope for voting on quotes.

This validation is introduced to ensure that users cannot vote on a
quote more than once.

Murph33 committed:
Add score method to tally up votes for a quote.

Also add a test that verifies that the quote.score method will count
tally votes associated with the quote, but not a vote that is not
associated with the quote.

Murph33 committed:
Add vote create action and tests for it

Murph33 committed:
Add vote update action and tests for it.

Murph33 committed:
Add vote destroy action and tests for it.

Murph33 committed:
Change controller and views to add voting.

Senjai commented on app/controllers/votes_controller.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def update
+    @quote = Quote.find params[:quote_id]
```
@quote = and @vote = at the beginning of every action can be extracted out to a before_action.

Murph33 committed:
Merge pull request #15 from StemboltHQ/add_votes

Add vote model along with validation test.

Murph33 commented on app/controllers/application_controller.rb:
```diff
@@ -2,4 +2,14 @@ class ApplicationController < ActionController::Base
+
+  helper_method :vote_types
+
+  def vote_types
```
This should be local to the controller that uses it, it does not need to be application wide.

I know later on you switch to the enum approach, we can use I18n at that point as discussed.

~Senjai

Murph33 committed:
Change order of input on form.

It makes more sense to order it as "Who said what where?" as opposed to
the current form of "Who was where when they said what?".

Commit:
Merge pull request #16 from StemboltHQ/change_quote_form_order

Change order of input on form.

Murph33 committed:
Change format of quote partial

Removed 'in the' from the partial since people can be quoted in
different states such as "while driving to work", "in the board room",
"on the phone", etc.  Provides more flexibility in the description of
how the quote was delivered.

Murph33 committed:
Add example to quote form to instruct format.

This provides an example of what the quote will be formatted like.
Think of Apples to Apples.

Commit:
Merge pull request #13 from StemboltHQ/add_instruction_to_form

Add example to quote form to instruct format.

Murph33 commented:
This should have been made as a PR to master.  Please see #17
